File: ch3/UVa11935.ml
open Printf
open Scanf

let eps = 1e-9

type event =
  | Fuel of float
  | Leak
  | Gas_station
  | Mechanic
  | Goal

let rec binsearch can lo hi =
  if hi-.lo < eps then
    hi
  else
    let mid = (lo+.hi) /. 2.0 in
    if can mid then
      binsearch can lo mid
    else
      binsearch can mid hi

let rec read_events () =
  let d, line = scanf "%f %[^\n]\n" (fun x y -> x, y) in
  let e = match line with
    | "Leak" -> Leak
    | "Gas station" -> Gas_station
    | "Mechanic" -> Mechanic
    | "Goal" -> Goal
    | _ ->
      let n = sscanf line "Fuel consumption %f" (fun x -> x) in
      if n = 0.0 then raise Exit
      else Fuel n
  in
  (d, e) :: match e with
  | Goal -> []
  | _ -> read_events ()

let can_events events original_f =
  let f = ref original_f in
  let cur_d = ref 0.0 in
  let cur_n = ref 0.0 in
  let leak_rate = ref 0.0 in
  let rec go = function
    | [] -> true
    | (d, e) :: events' ->
      let dt = d -. !cur_d in
      f := !f -. dt /. 100.0 *. !cur_n;
      f := !f -. dt *. !leak_rate;
      if !f < 0.0 then false
      else begin
        begin
          match e with
          | Fuel n -> cur_n := n
          | Leak -> leak_rate := !leak_rate +. 1.0
          | Gas_station -> f := original_f
          | Mechanic -> leak_rate := 0.0
          | Goal -> ()
        end;
        cur_d := d;
        go events'
      end
  in
  go events

let () =
  try
    while true do
      let events = read_events () in
      let ans = binsearch (can_events events) 0.0 10000.0 in
      printf "%.3f\n" ans;
    done
  with Exit -> ()

File: ch3/dnc/UVa11935.py
import sys

events = []

def can(f: float):
  global events

  cur_d, cur_n, leak_rate, dt = 0, 0, 0, 0
  original_f = f

  for i in range(len(events)):
    dt = events[i][0] - cur_d
    f -= dt / 100 * cur_n
    f -= dt * leak_rate
    if f < 0:
      return False
    if events[i][1] <= 0:
      cur_n = -events[i][1]
    elif events[i][1] == 1:
      leak_rate += 1
    elif events[i][1] == 2:
      f = original_f
    elif events[i][1] == 3:
      leak_rate = 0
    elif events[i][1] == 4:
      break
    cur_d = events[i][0]
  return f >= 0.0


if __name__ == '__main__':
  for line in sys.stdin:
    n = int(line.split()[-1])
    if n == 0:
      break
    events = [[0, -n]]

    while True:
      tkn = sys.stdin.readline().strip('\n').split()
      d = int(tkn[0])
      state = ' '.join(tkn[1:])
      if state[0:4] == 'Fuel':
        n = int(tkn[-1])
        events.append([d, -n])
      elif state == 'Leak':
        events.append([d, 1])
      elif state == 'Gas station':
        events.append([d, 2])
      elif state == 'Mechanic':
        events.append([d, 3])
      elif state == 'Goal':
        events.append([d, 4])
        break

    lo, hi = 0.0, 10000.0
    for i in range(50):
      mid = (lo+hi)/2
      if can(mid):
        hi = mid
      else:
        lo = mid

    print('%.3lf' % hi)
